File: crates/dewasm-backend-perl/units/rt/f32_demote.pl
# requires: rt/f32, rt/f32_bits, rt/f32_from_bits
# Demotion of a NaN yields the canonical NaN payload with the sign kept.
sub f32_demote {
    my $x = $_[0];
    if ($x != $x) {
        return Rt::f32_from_bits((Rt::f32_bits($x) & 0x80000000) | 0x7FC00000);
    }
    return Rt::f32($x);
}

File: crates/dewasm-backend-ruby/units/rt/f32_demote.rb
# requires: rt/f32, rt/f32_bits, rt/f32_from_bits
def f32_demote(x)
  x.nan? ? f32_from_bits((f32_bits(x) & 0x8000_0000) | 0x7fc0_0000) : f32(x)
end
